args/core/containers/sparse_map.hpp

```diff
@@ -420,14 +421,15 @@ namespace args::core
 				*itr_value = std::forward<value_type>(value_type());
 
 				auto itr_key = m_dense_key.begin() + m_size;
-				*itr_key = key;
-
-				m_sparse[key] = m_size;
+				*itr_key = std::move(key);
+				k = *itr_key;
+				m_sparse[k] = m_size;
 				++m_size;
 			}
+			else
+				k = key;
 
-
-			return m_dense_value[m_sparse[key]];
+			return m_dense_value[m_sparse[k]];
```

Comment:
mind explaining a bit what is going on here ?

Reply:
(in the form of comments preferably)

Reply:
uhm sure, but it's quite standard. using a move operation might invalidate the previous rvalue or lvalue ref you had. thus you cannot use that item anymore.

Reply:
yes, but that does not being quite standard does not mean it is particularly readable as a whole
```cpp
		value_reference operator[](key_type&& key)
		{
			key_type k;
			if (!contains(key))
			{
				if (m_size >= m_capacity)
					reserve(m_size + 1);

				auto itr_value = m_dense_value.begin() + m_size;
				*itr_value = std::forward<value_type>(value_type());

				auto itr_key = m_dense_key.begin() + m_size;
				*itr_key = std::move(key);
				k = *itr_key;
				m_sparse[k] = m_size;
				++m_size;
			}
			else
				k = key;

			return m_dense_value[m_sparse[k]];
		}
```
here is the piece of code in question without github diff stuff around it. Not super readbale imo.
It does not help that we have itr_key, key, and k here which are seamingly the same things (and all quite terrible names for variables)

Reply:
![Untitled](https://user-images.githubusercontent.com/50330170/89197873-ac4c6380-d5ac-11ea-9f13-9cf11c730d75.png)

is this okay?

Reply:
yes